Add BIO_s_dgram_pair

hlandau opened the pull request:
:tada: 

This is based on my sendmmsg/recvmmsg branch since it's dependent on that.

Supports multithreaded use.

mattcaswell commented on crypto/bio/bss_dgram_pair.c:
```diff
@@ -0,0 +1,1026 @@
+    if (CRYPTO_THREAD_write_lock(b->lock) == 0)
+        return -1;
+
+    for (i=0; i<num_msg; ++i) {
```
Missing spaces around `=` and `<`

mattcaswell commented on crypto/bio/bss_dgram_pair.c:
```diff
@@ -0,0 +1,1026 @@
+    if (CRYPTO_THREAD_write_lock(peerb->lock) == 0)
+        return -1;
+
+    for (i=0; i<num_msg; ++i) {
```
Missing spaces around `=` and `<`

mattcaswell commented on doc/man3/BIO_s_dgram_pair.pod:
```diff
@@ -0,0 +1,220 @@
+
+BIO_s_dgram_pair() returns the method for a BIO datagram pair. A BIO datagram
+pair is similar to a BIO pair (see BIO_s_bio()) but has datagram semantics.
+Broadly, this means that  the length of the buffer passed to a write call will
```
"that the" - too many spaces here

mattcaswell commented on doc/man3/BIO_s_dgram_pair.pod:
```diff
@@ -0,0 +1,220 @@
+whether they support certain capabilities; see CAPABILITY INDICATION.
+
+BIO_new_bio_dgram_pair() combines the calls to BIO_new(), BIO_make_bio_pair()
+and BIO_set_write_buf_size() to create a connected  pair of BIOs B<bio1>,
```
`BIO_new()`, `BIO_make_bio_pair()` and `BIO_set_write_buf_size()` are described on another man page so should be `L<BIO_new(3)>`, etc

mattcaswell commented on doc/man3/BIO_s_dgram_pair.pod:
```diff
@@ -0,0 +1,220 @@
+Each half of a BIO datagram pair can have capability flags set on it which
+indicate whether source and destination addresses will be honoured by the reader
+and whether they will be provided by the writer. These capability flags should
+be set via a call to BIO_dgram_set_caps(), and these capabilities wiill be
```
typo: wiill

mattcaswell commented on test/bio_dgram_test.c:
```diff
@@ -294,6 +295,282 @@ static int test_bio_dgram(void)
+                            | BIO_DGRAM_CAP_PROVIDES_DST_ADDR;
+
+    if (!TEST_int_eq(RAND_bytes((uint8_t *)key, sizeof(key)), 1))
+        goto err;
```
This test is problematic because it is not repeatable. If a bug is data dependent in some way then this test could fail sometimes and pass on other times. Can we make it repeatable somehow?

openssl-machine commented:
This PR is in a state where it requires action by @openssl/otc but the last update was 30 days ago

hlandau commented:
Rebased.

hlandau commented:
Rebased and updated now that the API is merged.

As with the other PR, unsure what's going on with doc-nits. This PR doesn't change the file it's complaining about.

t8m commented:
> Rebased and updated now that the API is merged.
> 
> As with the other PR, unsure what's going on with doc-nits. This PR doesn't change the file it's complaining about.

It's a recent regression, I am looking into it.

paulidale commented on crypto/bio/bss_dgram_pair.c:
```diff
@@ -0,0 +1,1132 @@
+            /* Restore original state. */
+            peerb->rbuf.idx[1] = saved_idx;
+            peerb->rbuf.count  = saved_count;
+            return -BIO_R_NON_FATAL;
```
Would introducing a BIO_R_TOO_SMALL error be more useful?  I'm not too fussed, but "non fatal" will cause questions.

hlandau replied:
`BIO_R_NON_FATAL` was added to act as a system independent equivalent of `EWOULDBLOCK`, basically. We could introduce another code but then we'd be changing the mmsg API yet again and updating BIO_err_is_non_fatal. Feels a bit weird to do so when this result code is brand new and so there are no conflicting uses. I thought about calling it `TRANSIENT` or `WOULD_BLOCK` rather than `NON_FATAL` but elected for parity with our existing BIO_err_is_non_fatal function...

paulidale commented on crypto/bio/bss_dgram_pair.c:
```diff
@@ -0,0 +1,1132 @@
+    struct bio_dgram_pair_st *x, *y;
+
+    x = (a->role == 1) ? a : b;
+    y = (a->role == 1) ? b : a;
```
Does the unlock order matter?  If not, it would be simplier to just unlock both directly.

I guess it's not overly important, it's just some register shuffling.

paulidale reviewed (commented):
This looks good.  Mostly nits but a couple of concerns.

hlandau commented on crypto/bio/bss_dgram_pair.c:
```diff
@@ -0,0 +1,1132 @@
+static size_t dgram_pair_read_inner(struct bio_dgram_pair_st *b, uint8_t *buf,
+                                    size_t sz);
+
+#define BIO_MSG_N(array, n) (*(BIO_MSG *)((char *)(array) + (n)*stride))
```
`stride` will always be `sizeof(BIO_MSG)` (the size it was when the calling code was compiled). So this should I believe be OK, unless we add a field to the structure which changes its alignment (since it contains a `uint64_t` already, this seems unlikely).

hlandau commented:
Updated.

mattcaswell commented on crypto/bio/bss_dgram_pair.c:
```diff
@@ -0,0 +1,1141 @@
+        return 0;
+
+    switch (cmd) {
+
```
Remove this blank line

mattcaswell commented on crypto/bio/bss_dgram_pair.c:
```diff
@@ -0,0 +1,1141 @@
+    y = (a->role == 1) ? b : a;
+
+    OPENSSL_assert(a->role != b->role);
+    OPENSSL_assert(a != b && x != y);
```
Don't use `OPENSSL_assert` inside libcrypto or libssl. Use `ossl_assert` - or just `assert` if you must.

mattcaswell commented on include/openssl/bioerr.h:
```diff
@@ -64,7 +66,5 @@
 # define BIO_R_WRITE_TO_READ_ONLY_BIO                     126
 # define BIO_R_WSASTARTUP                                 122
-# define BIO_R_LOCAL_ADDR_NOT_AVAILABLE                   148
-# define BIO_R_NON_FATAL                                  149
```
Why these deletions?

hlandau replied:
These are moved above

hlandau commented:
Updated.

openssl-machine commented:
24 hours has passed since 'approval: done' was set, but as this PR has been updated in that time the label 'approval: ready to merge' is not being automatically set.  Please review the updates and set the label manually.

paulidale commented:
Merged.

levitte commented on crypto/bio/bss_dgram_pair.c:
```diff
@@ -0,0 +1,1136 @@
+
+        ring_buf_pop(&b->rbuf, src_len);
+
+        buf         += src_len;
```
On line 743 above, there's a check `(buf != NULL)`.  It should probably be present here as well
@hlandau
